source/opt/constants.cpp

```diff
@@ -50,50 +40,52 @@ std::vector<const analysis::Constant*> ConstantManager::GetConstantsFromIds(
 }
 
 ir::Instruction* ConstantManager::BuildInstructionAndAddToModule(
-    std::unique_ptr<analysis::Constant> c, ir::Module::inst_iterator* pos) {
-  analysis::Constant* new_const = c.get();
+    const Constant* c, ir::Module::inst_iterator* pos) {
   uint32_t new_id = context()->TakeNextId();
-  const_val_to_id_[new_const] = new_id;
-  id_to_const_val_[new_id] = std::move(c);
-  auto new_inst = CreateInstruction(new_id, new_const);
-  if (!new_inst) return nullptr;
+  auto new_inst = CreateInstruction(new_id, c);
+  if (!new_inst) {
+    return nullptr;
+  }
   auto* new_inst_ptr = new_inst.get();
   *pos = pos->InsertBefore(std::move(new_inst));
   ++(*pos);
   context()->get_def_use_mgr()->AnalyzeInstDefUse(new_inst_ptr);
+  MapConstantToInst(c, new_inst_ptr);
   return new_inst_ptr;
 }
 
-analysis::Constant* ConstantManager::FindRecordedConstant(uint32_t id) const {
-  auto iter = id_to_const_val_.find(id);
-  if (iter == id_to_const_val_.end()) {
-    return nullptr;
+ir::Instruction* ConstantManager::GetDefiningInstruction(
+    const Constant* c, ir::Module::inst_iterator* pos) {
+  uint32_t decl_id = FindDeclaredConstant(c);
+  if (decl_id == 0) {
+    auto iter = context()->types_values_end();
+    if (pos == nullptr) pos = &iter;
+    return BuildInstructionAndAddToModule(c, pos);
   } else {
-    return iter->second.get();
+    return context()->get_def_use_mgr()->GetDef(decl_id);
   }
 }
 
-std::unique_ptr<analysis::Constant> ConstantManager::CreateConstant(
-    const analysis::Type* type,
-    const std::vector<uint32_t>& literal_words_or_ids) const {
-  std::unique_ptr<analysis::Constant> new_const;
+const Constant* ConstantManager::CreateConstant(
+    const Type* type, const std::vector<uint32_t>& literal_words_or_ids) const {
+  std::unique_ptr<Constant> new_const;
```

Comment:
Unused variable

Reply:
Fixed. Odd that the compiler didn't get it.